src/metabase/lib/filter/update.cljc

```diff
@@ -0,0 +1,179 @@
+  points.
+
+  There are three types of brush filters:
+
+  - [[update-temporal-filter]], which works on a single temporal column (e.g. zooming in on certain range in a
+    timeseries visualization)
+
+  - [[update-numeric-filter]], which works on a single numeric column
+
+  - [[update-lat-lon-filter]], which works on a latitude and longitude column pair. This is used with map visualizations --
+    draw a box between two points to zoom in to that part of the map.
+
+  If there is no existing filter on the column(s), these add a new filter. Existing filters are replaced."
+  (:require
+   [metabase.lib.breakout :as lib.breakout]
+   [metabase.lib.equality :as lib.equality]
+   [metabase.lib.filter :as lib.filter]
+   [metabase.lib.ref :as lib.ref]
+   [metabase.lib.remove-replace :as lib.remove-replace]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.literal :as lib.schema.literal]
+   [metabase.lib.schema.metadata :as lib.schema.metadata]
+   [metabase.lib.schema.temporal-bucketing :as lib.schema.temporal-bucketing]
+   [metabase.lib.temporal-bucket :as lib.temporal-bucket]
+   [metabase.lib.util :as lib.util]
+   [metabase.shared.util.time :as shared.ut]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]))
+
+(defn- is-ref-for-column? [expr column]
+  (and (lib.util/clause-of-type? expr :field)
+       (lib.equality/find-matching-column expr [column])))
+
+(mu/defn ^:private remove-existing-filters-against-column :- ::lib.schema/query
+  "Remove any existing filters clauses that use `column` as the first arg in a stage of a `query`."
+  [query        :- ::lib.schema/query
+   stage-number :- :int
+   column       :- ::lib.schema.metadata/column]
+  (reduce
+   (fn [query [_tag _opts expr :as filter-clause]]
+     (if (is-ref-for-column? expr column)
+       (lib.remove-replace/remove-clause query stage-number filter-clause)
+       query))
+   query
+   (lib.filter/filters query stage-number)))
+
+(mu/defn update-numeric-filter :- ::lib.schema/query
+  "Add or update a filter against `numeric-column`. Adapted from
+  https://github.com/metabase/metabase/blob/98bcd7fc3102bd7c07e8b68878c3738f3cb8727b/frontend/src/metabase-lib/queries/utils/actions.js#L151-L154"
+  ([query numeric-column start end]
+   (update-numeric-filter query -1 numeric-column start end))
+
+  ([query          :- ::lib.schema/query
+    stage-number   :- :int
+    numeric-column :- ::lib.schema.metadata/column
+    start          :- number?
+    end            :- number?]
+   (let [[start end] (sort [start end])]
+     (-> query
+         (remove-existing-filters-against-column stage-number numeric-column)
+         (lib.filter/filter stage-number (lib.filter/between numeric-column start end))))))
+
+(def ^:private temporal-filter-min-num-rows
+  "Minimum number of rows an updated query should return before; if it will return less than this, switch to
```

Comment:
```suggestion
  "Minimum number of rows an updated query should return; if it will return less than this, switch to
```
I don't understand what "before" is referring to.

I was a bit confused by the part about the "rows" returned by the query. It suggests that we actually run the/some query to see how many rows it returns. Reading further I understood this is about number of distinct values that the breakout can have given its unit.

Reply:
Fixed and added a lot more comments clarifying what I meant. I changed "rows" to "points" here since it's a better reflection